src/core/client_config/resolvers/zookeeper_resolver.c

```diff
@@ -0,0 +1,452 @@
+    grpc_client_config_set_lb_policy(config, lb_policy);
+    GRPC_LB_POLICY_UNREF(lb_policy, "construction");
+    grpc_resolved_addresses_destroy(addresses);
+    gpr_free(subchannels);
+  }
+  gpr_mu_lock(&r->mu);
+  GPR_ASSERT(r->resolving);
+  r->resolving = 0;
+  if (r->resolved_config) {
+    grpc_client_config_unref(r->resolved_config);
+  }
+  r->resolved_config = config;
+  r->resolved_version++;
+  zookeeper_maybe_finish_next_locked(r);
+  gpr_mu_unlock(&r->mu);
+
+  GRPC_RESOLVER_UNREF(&r->base, "zookeeper-resolving");
+}
+
+/* Callback function for each DNS resolved address */
+static void zookeeper_dns_resolved(void *arg, grpc_resolved_addresses *addresses) {
+  size_t i;
+  zookeeper_resolver *r = arg;
+  gpr_mu_lock(&r->mu);
+  r->resolved_num++;
+  r->resolved_addrs->addrs = gpr_realloc(r->resolved_addrs->addrs, 
+                            sizeof(grpc_resolved_address) * (r->resolved_addrs->naddrs + addresses->naddrs));
+  for (i = 0; i < addresses->naddrs; i++) {
+    memcpy(r->resolved_addrs->addrs[i + r->resolved_addrs->naddrs].addr, 
+          addresses->addrs[i].addr, addresses->addrs[i].len);
+    r->resolved_addrs->addrs[i + r->resolved_addrs->naddrs].len = addresses->addrs[i].len;
+  }
+
+  r->resolved_addrs->naddrs += addresses->naddrs;
+  grpc_resolved_addresses_destroy(addresses);
+
+  /* Wait for all addresses to be resolved */
+  if (r->resolved_num == r->resolved_total) {
+    gpr_mu_unlock(&r->mu);
+    zookeeper_on_resolved(r, r->resolved_addrs);
+  } else {
+    gpr_mu_unlock(&r->mu);
+  }
+}
+
+/* Parse json format address of a zookeeper node */
+static char *zookeeper_parse_address(char *buffer, int buffer_len) {
+  char *host;
+  char *port;
+  char *address;
+  grpc_json *json;
+  grpc_json *cur;  
+
+  address = NULL;
+  json = grpc_json_parse_string_with_len(buffer, buffer_len);
+  if (json != NULL) {
+    host = NULL;
+    port = NULL;
+    for (cur = json->child; cur != NULL; cur = cur->next) {
+      if (!strcmp(cur->key, "host")) {
+        host = (char *)cur->value;
+        if (port != NULL)
+          break;
+      }
+      else if (!strcmp(cur->key, "port")) {
+        port = (char *)cur->value;
+        if (host != NULL)
+          break;
+      }
+    }
+    if (host != NULL && port != NULL) {
+      address = gpr_malloc(GRPC_MAX_SOCKADDR_SIZE);
+      memset(address, 0, GRPC_MAX_SOCKADDR_SIZE);
+      strcat(address, host);
+      strcat(address, ":");
+      strcat(address, port);
+    }
+    grpc_json_destroy(json);
+  } 
+
+  return address;
+}
+
+static void zookeeper_get_children_node_completion(int rc, const char *value, int value_len, 
+                                          const struct Stat *stat, const void *arg) {
+  char *address = NULL;
+  zookeeper_resolver *r = (zookeeper_resolver *)arg;
+
+  if (rc) {
+    gpr_log(GPR_ERROR, "Error in getting a child node of %s", r->name);
+    return;
+  }
+
+  address = zookeeper_parse_address((char *)value, value_len);
```

Comment:
Try to avoid using casting to remove const as much as possible: you are braking a promise you made on the function signature (namely, that the contents of "value" wouldn't be modified). In this case it's a "private" (static) function, but if this were a public function, that'd be potentially very bad.

It's better to change the function signature to ...char* value... to avoid lying about the constness.